Source/WebCore/editing/cocoa/HTMLConverter.mm

```diff
@@ -593,7 +593,7 @@ static bool stringFromCSSValue(CSSValue& value, String& result)
 
     if (RefPtr<CSSValue> value = inlineStylePropertyForElement(element, propertyId)) {
         String result;
-        if (value->isInheritValue())
+        if (isValueID(*value, CSSValueInherit))
```

Comment:
We could do that, but I thought there was a chance that we might optimize `isValueID` differently at some point, so was leaving it in place for now. It would be a quick mechanical process to remove it, could do it with global replace. I choose not to add to this patch, though.

I think we might want change `CSSValue` to represent value IDs differently. Perhaps just encode the bits right in the `CSSValue` so we can fetch it without checking the type, putting the value `CSSValueInvalid` in for all CSS values that aren’t `CSSPrimitiveValue` containing a value ID. Another possibility would be to encode both the CSS value class and the value ID in a single 16-bit integer; all the value IDs would imply the primitive type. In that case, `isValueID` might be more efficient because it’s comparing with a type that it knows is not `CSSValueInvalid`, whereas `valueID` needs to check and return `CSSValueInvalid`. But also the entire `CSSValue` representation is really left over from when it was the CSS OM object, but that is no longer the case. I wonder why we are reference counting objects and allocating on the heap when so many are just holding an integer.

I also am unclear on why we are using so many different classes that are specific to various types. It seems we need a flexible way to hold the data, a basic object model. The property-specific code can know the shape of the data rather than having to use a property-specific value holder object. We might want to use a scalar-style class like `Color` instead of `Ref<CSSValue>` and do the reference counting internally. It’s great to share the reference counted bit, but not great to allocate on the heap just so we can do polymorphism.

These thoughts are not yet fully formed, but I suspect they could drastically simplify some of our code. My lack of familiarity with the new typed object model is holding me back a little, though.

Not clear what direction we should go, but I do think we might be able to improve efficiency with a better representation.

Reply:
That's fair.